### PR TITLE
feat: Add NAT policy specification

### PR DESCRIPTION
**Pull request by kklimonda-cl**
Create manual NAT policy specification based on the API schema.
Additionally, update codegen to better map schema to terraform:
- change terraform type of "nil" parameters from string to object: this affects this spec and ethernet (where one of variants is empty ha option). It's more in line with other variants this way.
- Make "nil" variants optional.

Making "nil" variants optional lets us skip variant if "none" variant would be selected.

### CODE REVIEW

File: specs/policies/network-address-translation.yaml [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+                            - name: interface
+                              type: string
+                              profiles:
+                                - xpath: [interface]
+                              validators:
+                                - type: length
+                                  spec:
+                                    max: 31
+                            - name: ip
+                              type: string
```
Comment (michalbil) [inline specs/policies/network-address-translation.yaml:293]: I think this one should be an object and have `spec.variants.{ip|floating_ip} of type string.

File: specs/policies/network-address-translation.yaml [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+        - type: regexp
+          spec:
+            expr: "[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}"
+        - type: length
+          spec:
+            min: 3
+            max: 36
+      profiles:
+        - xpath: ["uuid"]
+    - name: from
```
[the remark this answers is not in the document]
Comment (kklimonda-cl) [reply specs/policies/network-address-translation.yaml:146]: Done, renamed `from` => `source_zones`, `to` => `destination_zone`, `source` => `source_addresses` and `destination` => `destination_addresses`

File: specs/policies/network-address-translation.yaml (hunk 1 above)
Comment (kklimonda-cl) [reply specs/policies/network-address-translation.yaml:364]: We've discussed it during todays meeting: in XML API both ip and floating-ip are on the same level as interface, not wrapped in any other parent element. We've discussed adding this wrapping to terraform, but decided that wrapped version[1] would not be more readable.

[1] Wrapped versions would be one of:
```
destination_translation = {
  ip = {
    ip = "value"
    floating-ip = "or value"
  }
  interface = "ethernet1/1"
}
```
destination_translation = {
  ip = { # wrapper name is example, but couldn't figure out a better one during a meeting
    type = "ip" # or floating-ip
    value = "ip-value"
  }
  interface = "ethernet1/1"
}
```

File: assets/terraform/test/resource_nat_policy_test.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+	"github.com/hashicorp/terraform-plugin-testing/statecheck"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+	"github.com/hashicorp/terraform-plugin-testing/tfjsonpath"
+)
+
+type deviceType int
+
+const (
+	devicePanorama deviceType = iota
+	deviceFirewall deviceType = iota
```
Comment (jaspalo) [inline assets/terraform/test/resource_nat_policy_test.go:29]: note: `iota` can be skipped

File: assets/terraform/test/resource_nat_policy_test.go [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+		for _, elt := range rules {
+			result = append(result, config.StringVariable(prefixed(elt)))
+		}
+
+		return result
+	}
+
+	device := devicePanorama
+
+	location := locationByDeviceType(device)
```
Comment (jaspalo) [inline assets/terraform/test/resource_nat_policy_test.go:163]: suggestion: we could benefit from explicit initialisation of the location in the context of the test. The cases don't switch a location back and forth, so the function doesn't really help much.

File: assets/terraform/test/resource_nat_policy_test.go [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+		ProtoV6ProviderFactories: testAccProviders,
+		CheckDestroy:             natPolicyCheckDestroy(prefix, location),
+		Steps: []resource.TestStep{
+			{
+				Config: makeConfig(prefix, device),
+				ConfigVariables: map[string]config.Variable{
+					"rule_names": config.ListVariable(withPrefix(rulesInitial)...),
+				},
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownValue(
```
Comment (jaspalo) [inline assets/terraform/test/resource_nat_policy_test.go:180]: suggestion: state and plan checks can be prepared based on the rule test data.

File: assets/terraform/test/panos_sweeper_test.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package provider_test
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+)
+
```
[the remark this answers is not in the document]
Comment (kklimonda-cl) [reply assets/terraform/test/panos_sweeper_test.go:8]: Not sure what do you mean here - is this about TestMain function?

File: assets/terraform/test/resource_nat_policy_test.go [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+				NgfwDevice: "localhost.localdomain",
+				Vsys:       "vsys1",
+			},
+		}
+	}
+
+	return location
+}
+
+func natPolicyPreCheck(prefix string, location nat.Location) {
```
[the remark this answers is not in the document]
Comment (kklimonda-cl) [reply assets/terraform/test/resource_nat_policy_test.go:337]: This precheck seeds remote server to verify that entries are deleted when terraform changes are applied.

File: assets/terraform/test/resource_nat_policy_test.go (hunk 4 above)
Comment (kklimonda-cl) [reply assets/terraform/test/resource_nat_policy_test.go:163]: We need some way to run tests against panorama and VM series, which have different locations.

File: assets/terraform/test/resource_nat_policy_test.go [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+				},
+			},
+		},
+	})
+}
+
+const configTmpl = `
+variable "rule_names" { type = list(string) }
+
+resource "panos_nat_policy" "{{ .ResourceName }}" {
```
[the remark this answers is not in the document]
Comment (kklimonda-cl) [reply assets/terraform/test/resource_nat_policy_test.go:260]: I agree on using config variables, that makes sense.
I'd agree that using simple string formatting would be easier, if not for the fact that sometimes we'll want to use string formatting (as in this case) and then we'd have to use "%%s" to escape it in the code.

File: assets/terraform/test/resource_nat_policy_test.go (hunk 5 above)
Comment (kklimonda-cl) [reply assets/terraform/test/resource_nat_policy_test.go]: When testing I prefer to be more explicit about the expected state - I've added helper functions to build state and plan checks in a more streamlined manner.

**Review by jaspalo — APPROVED** [review]

**Pull request merged**